src/body_writer.rs

```diff
@@ -0,0 +1,349 @@
+        self.emit_finished_chunk().await
+    }
+
+    /// Append to the buffer
+    fn append_current_chunk(&mut self, buf: &[u8]) -> usize {
+        let buffered = usize::min(buf.len(), self.buf.len() - NEWLINE.len() - self.pos);
+        if buffered > 0 {
+            self.buf[self.pos..self.pos + buffered].copy_from_slice(&buf[..buffered]);
+            self.pos += buffered;
+        }
+        buffered
+    }
+
+    /// Finish the current chunk by writing the header
+    fn finish_current_chunk(&mut self) {
+        // Write the header in the allocated position position
+        let chunk_len = self.pos - self.header_pos - self.allocated_header;
+        let header_buf = &mut self.buf[self.header_pos..self.header_pos + self.allocated_header];
+        let header_len = write_chunked_header(header_buf, chunk_len);
+
+        // Move the payload if the header length was not as large as it could possibly be
+        let spacing = self.allocated_header - header_len;
+        if spacing > 0 {
+            self.buf.copy_within(
+                self.header_pos + self.allocated_header..self.pos,
+                self.header_pos + header_len,
+            );
+            self.pos -= spacing
+        }
+
+        // Write newline footer after chunk payload
+        self.buf[self.pos..self.pos + NEWLINE.len()].copy_from_slice(NEWLINE);
+        self.pos += 2;
+
+        self.header_pos = self.pos;
+        self.allocated_header = get_max_chunk_header_size(self.buf.len() - self.header_pos);
+        self.pos = self.header_pos + self.allocated_header;
+    }
+
+    async fn emit_finished_chunk(&mut self) -> Result<(), C::Error> {
+        self.conn.write_all(&self.buf[..self.header_pos]).await?;
+        self.header_pos = 0;
+        self.allocated_header = get_max_chunk_header_size(self.buf.len());
+        self.pos = self.allocated_header;
+        Ok(())
+    }
+}
+
+impl<C> ErrorType for BufferingChunkedBodyWriter<'_, C>
+where
+    C: Write,
+{
+    type Error = embedded_io::ErrorKind;
+}
+
+impl<C> Write for BufferingChunkedBodyWriter<'_, C>
+where
+    C: Write,
+{
+    async fn write(&mut self, buf: &[u8]) -> Result<usize, Self::Error> {
+        let written = self.append_current_chunk(buf);
+        if written < buf.len() {
+            self.finish_current_chunk();
+            self.emit_finished_chunk().await.map_err(|e| e.kind())?;
+        }
+        Ok(written)
+    }
+
+    async fn flush(&mut self) -> Result<(), Self::Error> {
+        if self.header_pos > 0 {
+            self.finish_current_chunk();
+            self.emit_finished_chunk().await.map_err(|e| e.kind())?;
+        }
+        self.conn.flush().await.map_err(|e| e.kind())
+    }
+}
+
+const fn get_hex_chars(number: usize) -> u32 {
+    if number == 0 {
+        1
+    } else {
+        (usize::BITS - number.leading_zeros()).div_ceil(4)
+    }
+}
+
+const fn get_max_chunk_header_size(buffer_size: usize) -> usize {
+    if buffer_size >= NEWLINE.len() + NEWLINE.len() {
```

Comment:
Do you have something against `2*`? :)

Also this function would look much simpler by using `checked_sub()`

Reply:
I guess I do :-D No of cause not! I have aligned, thanks!